### PR TITLE
Propagate TLS handshake exception in `SslCompletionEvent`

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

- `SslCompletionEvent` can have a cause if a TLS handshake is completed exceptionally. As the cause isn't taken into account, users could see the root cause of the failure.

- A legacy HTTPS server such as Microsoft-IIS/8.5 resets the connection if no cipher suites in common without respecting the handshake process. No hint is given to the handshake failure, users don't know why Armeria can't communicate with the server.

Modifications:

- If `SslCompletionEvent` is not successful, propagate the cause to the session promise.
- If an exception is raised during a handshake, create `IllegalStateException` with a hint and add it as a suppressed exception.

Result:

Exceptions that occurred during a TLS handshake are properly propagated to users.

### CODE REVIEW

**Comment by tobias-** [comment]
❤️

File: core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java [hunk 1]
```diff
@@ -489,11 +503,20 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
     private static Throwable getPendingException(ChannelHandlerContext ctx) {
+        final Throwable pendingException = getPendingException(ctx, null);
+        if (pendingException != null) {
+            return pendingException;
+        }
+        return ClosedSessionException.get();
+    }
+
+    @Nullable
+    private static Throwable getPendingException(ChannelHandlerContext ctx, @Nullable Throwable defaultCause) {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java:514]: It seems like `defaultCause` is always `null`.

**Review by minwoox — COMMENTED** [review]
Just one nit. 👍

File: core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java [hunk 2]
```diff
@@ -263,6 +264,17 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
...
             @Override
             public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+                if (handshakeFailed == null) {
+                    // An exception was raised before the handshake event was completed.
+                    // A legacy HTTPS server such as Microsoft-IIS/8.5 may reset the connection
+                    // if no cipher suites in common.
+                    final IllegalStateException maybeHandshakeException = new IllegalStateException(
+                            "An unexpected exception during TLS handshake. One possible reason behind the " +
+                            "failure is no cipher suites in common. " +
+                            "ciphers: " + sslCtx.cipherSuites(), cause);
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java:274]: How about this: (Not a complete diff :sweat_smile:)
```suggestion
                    final IllegalStateException maybeHandshakeException = new IllegalStateException(
                            "An unexpected exception during TLS handshake. Possible reasons: no cipher suites in common, unsupported TLS version, etc. " +
                            "(TLS version: " ... ", cipher suites: " + sslCtx.cipherSuites() + ')', cause);
```

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java:274]: Looks much better!  🙇‍♂️

File: core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java [hunk 3]
```diff
@@ -268,10 +268,12 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
                     // An exception was raised before the handshake event was completed.
                     // A legacy HTTPS server such as Microsoft-IIS/8.5 may reset the connection
                     // if no cipher suites in common.
+                    final String tlsVersion = sslHandler.engine().getSession().getProtocol();
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/client/HttpClientPipelineConfigurator.java:271]: Note: I checked if there's a case where `getSession()` or `getProtocol()` returns `null`, but it seems like JDK and OpenSSL implementation always return non-null.

**Review by trustin — APPROVED** [review]
Thanks, @ikhoon for the fix and @tobias- for reporting this issue. :bow:

**Review by minwoox — APPROVED** [review]
Thanks! 👍 👍 👍

**Comment by ikhoon** [comment]
The test seems flaky. The two events in `HttpClientPipelineConfigurator` and `HttpSessionHandler` may change depending on the operating system or somehow.

I marked it as a flaky test. Do you think the failure should be fixed, I will come up with another idea.

File: core/src/test/java/com/linecorp/armeria/client/IisServerCompatibilityTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    @RegisterExtension
+    static ServerExtension server = new ServerExtension() {
+        @Override
+        protected void configure(ServerBuilder sb) {
+            sb.https(0);
+            sb.tlsSelfSigned();
+            sb.childChannelPipelineCustomizer(childChannel -> {
+                // Simulate the behavior of IIS.
+                childChannel.channel().close();
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/client/IisServerCompatibilityTest.java:45]: I failed to reproduce the problem when I used different cipher suites. The TLS handshake was completed with the correct cause, and the connection was closed correctly.
 
IIS seems to violate the specification.

[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/client/IisServerCompatibilityTest.java:45]: To exactly simulate the behavior of IIS, we have to reset the connection when Client Hello message is received. I'm not sure we can catch the event with `SslHandler`.

File: core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java [hunk 5]
```diff
@@ -410,9 +409,29 @@ public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exc
...
+                // Expected event
+            } else {
+                Throwable handshakeException = sslCompletionEvent.cause();
+                final Throwable pendingException = getPendingException(ctx);
+                if (pendingException != null) {
+                    pendingException.addSuppressed(handshakeException);
+                    // Propagate suppressed exceptions in handshakeException to pendingException.
+                    for (Throwable suppressed : handshakeException.getSuppressed()) {
+                        pendingException.addSuppressed(suppressed);
+                    }
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/client/HttpSessionHandler.java:424]: First, I thought your suggestion worked. While testing, it seemed to work as expected. I might be confused while fixing multiple places.

Let me remove the lines.

File: core/src/test/java/com/linecorp/armeria/client/IisServerCompatibilityTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrhee17) [reply core/src/test/java/com/linecorp/armeria/client/IisServerCompatibilityTest.java:45]: I see 😅    What do you think of just removing the test? (or manually triggering a `SslCompletionEvent `)

Even if it is marked as a flaky test, it seems to fail too often

[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/test/java/com/linecorp/armeria/client/IisServerCompatibilityTest.java:45]: Agreed. I am thinking to add a `ChannelHandler` and try to close the connection when the first message is received. 

If it does not work, I will come up with another idea on that.

**Review by jrhee17 — APPROVED** [review]
Thanks @ikhoon ! 👍 🙇 👍   I'm happy as long as tests are passing 😄

**Pull request merged**